CellsUsed refactoring and fix

Pankraty opened the pull request:
Fixes #1026 

I started this branch right after discovering the issue with CellsUsed method during #968 fix but then had to switch to other activities.

To do things right I started a large refactoring
* I added a flagged enum `XLCellsUsedOptions` similar to `XLClearOptions` (I'd prefer to keep them separate even though they have same values)
* I heavily refactored `CellsUsed`, `FirstCellUsed`, `LastCellUsed` and other methods to consider merged ranges, conditional formats, data validation rules, etc.
* I marked all methods using `includeFormat` flag as `[Obsolete]`. After a couple versions we can remove them I suppose.
* ~I found and fixed another bug, which perhaps I should address separately. When cells are merged they didn't keep their own styles and values instead of acquiring the style of the first cell in the range and clearing everything else (contents, data validations, conditional formats...)~ Moved to #1030

Pankraty committed:
Add failing test for CellsUsed method

Pankraty committed:
Fix issues with CellsUsed logic

Pankraty committed:
Refactor classes using includeFormat flag to use XLCellsUsedOptions instead.

igitur commented on ClosedXML/Excel/Cells/XLCell.cs:
```diff
@@ -1991,9 +1991,9 @@ private Boolean SetRange(Object rangeObject)
-                    Worksheet.Range(_rowNumber, _columnNumber, maxRows, maxColumns).Clear();
+                    var maxRow = _rowNumber + asRange.RowCount() - 1;
+                    var maxColumn = _columnNumber + asRange.ColumnCount() - 1;
+                    Worksheet.Range(_rowNumber, _columnNumber, maxRow, maxColumn).Clear();
```
This was fixed in https://github.com/ClosedXML/ClosedXML/pull/1022/commits/2cc52eba90992671e4d396f3b12a7207b9f7f5fe , right?

Pankraty replied:
I noticed that too. I am going to rebase anyway, so this will be replaced with your solution.

igitur commented on ClosedXML/Excel/Cells/XLCells.cs:
```diff
@@ -24,15 +24,133 @@ public XLCells(bool usedCellsOnly, bool includeFormats, Func<IXLCell, Boolean> p
+            if (predicate == null)
+                _predicate = c => true;
+            else
+                _predicate = predicate;
```
`_predicate = predicate ?? _ => true;`

Pankraty replied:
Of course

Pankraty commented:
As for the bug with the merged ranges, I found that the issue is more complicated than I thought.
If I unmerge a range in the TransposeRange.xlsx I get the following result:
![image](https://user-images.githubusercontent.com/19576939/46885542-5784a980-ce69-11e8-874f-7a8d5c736dcb.png)

Both the "old" way and the "new" way produce incorrect result. Guess I'll extract this into a separate issue.

igitur commented:
Please see review comments in #1030

igitur commented:
Had a chance to look at this further? It just needs history rewriting. I'd like release soon.

Pankraty commented:
Will do it right away

Pankraty committed:
* Make options argument optional and includeFormats non-optional

* Replace obsolete methods in tests with the new ones
* Dispose workbook in the test
* Define usedCellsCandidates as IEnumerable rather than List

Pankraty commented:
I've rebased and squashed the branch and eliminated most of the issues. But more refactoring is still needed in order to properly handle large ranges in conditional formats and data validation rules (your recent test CFShiftedTruncateRange now takes forever to execute) :(

Unfortunately, I don't have enough time for this today. Will try to do it during the week.

igitur committed:
Ignore data validation and conditional format flags to improve performance of range insertion. DV and CF are handled elsewhere.

Pankraty committed:
Refactor FirstCellUsed and LastCellUsed in order to not hanging on very large conditional format rules, data validation rules or merged ranges.

igitur commented:
Methods made obsolete in this PR can be removed now, right?
